### DISCUSSION TITLE
Switch to a more recent Python version

### DISCUSSION THREAD

**Discussion by roland-ruedenauer** [zim-desktop-wiki/zim-desktop-wiki#1776] [Development]
Are there any plans for the project to switch to a more recent Python version?

Zim currently keeps compatibility to Python 3.2, which has been released already more than 10 years ago.
The application isn't small and I think it would really help having a least type hinting available in development, which would require moving on to at least Python 3.5. Python f-strings and new formatting would also be nice to have.

What do you think?

**Comment by jaap-karssenberg** [comment]
Sorry, but this is a bit theoretical. Haven't seen any substantial patch
where this version is limiting.

Op za 4 dec. 2021 16:06 schreef Roland Rüdenauer ***@***.***>:

**Comment by jaap-karssenberg** [comment]
PS. Also last time I broke this, I got big reports from users still on
older versions. So seems relevant to stay compatible unless there is a
really good reason.

Op za 4 dec. 2021 16:49 schreef Jaap Karssenberg ***@***.***

**Comment by roland-ruedenauer** [comment]
Well, I didn't start this discussion because I am bored. My experience from attempting to contribute is the reason I raised the question. I am pretty sure, you as founder of the project know the code base good enough. But to me type hints are really of great help, when I jump into the IDE after having paused for a while. I simply don't have to guess argument and return types that often.
In fact, I temporary add type hints and remove them before commiting, which is a silly thing.
As for staying compatible - yes, to some exted it surely is a good thing. But otherwise, the handful of people staying on rather outdated environments also are not forced to update their Zim application to newer versions.
But it really is up to you to decide about the direction. I was just interested if there is a plan to move or not.

**Reply by introt** [reply]
Type hints would be slightly less of an issue if the doc situation was better, but that's another thing. I'm a fan of Zim not requiring the latest and greatest libraries as it makes usage and development on eg. Debian stale [sic] possible, but I'd vouch for requiring at least the oldest officially supported version ([3.6 EOLs this month](https://devguide.python.org/#status-of-python-branches)). Correct me if I'm wrong, but in this day and age with containers and NIX I can't come up with a reason to run Zim using a vulnerable EOL version of Python.

**Reply by introt** [reply]
~Per 4ff6d4bb432153ecbf51ca8993ba08501118741a Ubuntu 16.04 Xenial Xerus (which also EOLs soon) is stuck on 3.5, which includes type hints; we could stop there for the time being.~ nevermind, 16.04 EOL'd in April (guess I'm still living in 2020)

**Reply by introt** [reply]
> In fact, I temporary add type hints and remove them before commiting

Have you been doing this manually or using some tool? We're on 3.6 now so you're more than welcome to get some annotations merged :)

**Reply by roland-ruedenauer** [reply]
Thanks for your efforts to make the switch to Python 3.6 happen.
I didn't use a tool to remove type annotations. And looking at my old (still open) PR, it seems I did not even remove all of them.
At least I don't have to worry about this anymore now.

**Reply by introt** [reply]
You're welcome, though I literally couldn't have done it without you - I wouldn't have had known 3.6 was on its way out unless you had prompted me to check.
I actually meant to ask you whether you used a tool for _adding_ the annotations - `mypy` suggests two and there are plenty of other options, so any pointers would be welcome! Need to manually remove docstrings that contain nothing but the type(s) though.
Seems Jaap's busy with refactoring the parse tree - could be a while before new feature PRs get reviewed, so I wouldn't hold my breath for the time being. Wish we had some traffic light type indicator for our BDFL ;)

**Comment by jaap-karssenberg** [comment]
Probably 3.5 is safe enough than.

And yes, I'm open to patches improving code documentation. Not sure you're
hubby are needed for that perse but not against them.

I think to improve code quality for contributors the biggest bang would be
refactoring the page view module (which is in my list) and improving test
coverage. The are a number of big test cases that need to be broken down to
be understandable.

Regards,

Jaap

Op zo 5 dec. 2021 13:37 schreef introt ***@***.***>:

**Reply by introt** [reply]
Cool!

Speaking of docs, what's your stance on Pydoctor[1] vs. Sphinx[2]? I reached a local maximum with my loopy approach (used for [2]) and ran out of time while (re)writing a recursive parser, but concluded that it could make more sense to adopt/adapt an existing tool – Pydoctor can read the docstrings as-is, but doesn't support custom fields (eg. `@emits` & `@signal`), while all the converters are lacking in one way or the other. I know online docs aren't a must for everyone, but personally I've found them quite helpful.

1. https://introt.github.io/zim-desktop-wiki/docs/api/
2. https://introt.github.io/zim-desktop-wiki/docs/build/html/index.html

**Comment by jaap-karssenberg** [comment]
BTW. What I would love is it there are type hints in the code that they are
enforced (and this checked for correctness) when running the test suite. Or
is that a bad idea?

Op zo 5 dec. 2021 13:44 schreef Jaap Karssenberg ***@***.***

**Reply by introt** [reply]
My older comment was misinformed; anyhow, [mypy](https://mypy.readthedocs.io/en/stable/existing_code.html) should fit the bill. Could get slow fast though, probs shouldn't be included in default suite (at least without caching)

**Reply by introt** [reply]
Latest version of mypy drops support for Python 3.5, though the version should be pinned anyway. Ubuntu 16.04 EOL'd last April but since people were using it in September I'm not sure if we can bump to 3.6 just yet...

**Reply by roland-ruedenauer** [reply]
Having static type checks enforced for pull requests using mypy is no rarity nowadays. And many IDEs already validate types while developing new code. It also makes sense having this type of validation enabled next to writing unit test.

Regarding type checking at runtime, I must admit to never having seen this at work. A quick search though revealed tools exist, that already support this. A stackoverflow [article](https://stackoverflow.com/questions/63343360/dynamic-checking-of-type-hints-in-python-3-5) mentions [pydantic](https://pydantic-docs.helpmanual.io/) and [typen](https://pypi.org/project/typen/). It would be interesting to play with this a bit and see what one could achieve with it.

In the future, I think flatpak / snap / appimage will solve the problem supporting older systems. Currently it doesn't seem to be the case, as e.g. flatpack is not available on Ubuntu 16.04 (provided I looked closely enough).

**Comment by jaap-karssenberg** [comment]
To be clear - I wouldn't want the production code running static typed all
the time. My main concern is that if there are annotations that we also
automatically check whether these are correct. Enforcing them in the test
suite could be a way to do that checking. But if it makes things overly
difficult let's not go crazy ...

**Comment by jaap-karssenberg** [comment]
Yes, one more patch to finish on my end and then I cycle back to checking
issues and pull requests. Afraid my schedule is so fluid I don't know
myself when that will be exactly.

Regards,

Jaap

Op ma 10 jan. 2022 09:46 schreef introt ***@***.***>:
